Improve fetch from multiple remotes

mateon1 opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the commits rather than adding commits on top. Use force-push when
pushing the updated commits (`jj git push` does that automatically when you
rewrite commits). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->
See #4920 for rationale.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes
- - [x] I have updated existing tests affected by this change

yuja commented:
Perhaps, the easiest workaround is to do `rebase_descendants()` per remote? Does this patch solve the problem?
```diff
diff --git a/cli/src/git_util.rs b/cli/src/git_util.rs
index 74ebe2d2fa..8949218232 100644
--- a/cli/src/git_util.rs
+++ b/cli/src/git_util.rs
@@ -496,6 +496,11 @@
             _ => user_error(err),
         })?;
         print_git_import_stats(ui, tx.repo(), &stats.import_stats, true)?;
+        let settings = tx.settings().clone(); // XXX silly clone
+        let num_rebased = tx.repo_mut().rebase_descendants(&settings)?;
+        if num_rebased > 0 {
+            writeln!(ui.status(), "Rebased {num_rebased} descendant commits")?;
+        }
     }
     warn_if_branches_not_found(
         ui,
```

It might be cleaner to do import at the end, but it's weird that `git:fetch()` accepts multiple remotes (and returns one `default_branch`.)

mateon1 commented:
Yuja's patch does work, with this result:
```
[mateon@ceres:/tmp/a]$ jj git fetch --all-remotes 
bookmark: pr@b [deleted] untracked
Abandoned 1 commits that are no longer reachable.
bookmark: master@c [updated] untracked

[mateon@ceres:/tmp/a]$ jj op show 63915faff2bc
63915faff2bc mateon@ceres 31 seconds ago, lasted 11 milliseconds
fetch from git remote(s) b,c
args: jj git fetch --all-remotes

Changed commits:
○  Change xqknuwottovp
   + xqknuwot bdd92bb1 master@c | (empty) Merge remote-tracking branch 'b/pr'

Changed remote bookmarks:
master@c:
+ untracked xqknuwot bdd92bb1 master@c | (empty) Merge remote-tracking branch 'b/pr'
- untracked qrmnvrwp 8ac29516 other
pr@b:
+ untracked (absent)
- untracked mszwrokv 1780e38a work
```

So it's a direct fix to the current bug while keeping existing behavior.
Personally I find it rather unintuitive that branch changes on one remote can mess with your local work even if the commits still exist in the shared history.
Although frankly I find it surprising that any kind of fetch can rebase your local commits - my expectation as a new jujutsu user was having to rebase your work yourself when a fetch makes your commits stale. I also expected fetch from multiple remotes to be significantly more atomic than it is now.

Re: default_branch, that's also a bit weird. It's only used from jj git clone, which only supports a single remote. Even if it did support multiple, how could you possibly know which remote is 'most authoritative' for choosing the default branch? What if you clone two forks of the same repo? What if you clone two completely independent repos?
I'm not sure what the most correct thing to do here would be. Returning a Vec of default_branches corresponding to the passed remotes? Making a 'clone' implementation that supports only one remote, handles default_branch, and ripping out that code from fetch?

essiene commented:
> > I find it surprising that any kind of fetch can rebase your local commits - my expectation as a new jujutsu user was having to rebase your work yourself when a fetch makes your commits stale.
> 
> You can turn it off by `git.abandon-unreachable-commits = false`, but you'll need to clean up old branches manually (if the remote does rebase/squash merges.)
> 
> > Re: default_branch, that's also a bit weird. It's only used from jj git clone, which only supports a single remote. Even if it did support multiple, how could you possibly know which remote is 'most authoritative' for choosing the default branch?
> 
> Perhaps, the API can be refactored to something like this:
> 
> ```rust
> let mut peer = GitFetch::new(..);
> let default_branch = peer.fetch(remote_name, branch_names, ..);
> ...
> let stats = peer.import_refs();
> ```


This seems like a good approach.

We could call peer.fetch() in a loop for all remotes and at the end, call peer.import_refs just once.
> 
> (and `git::fetch()` will be a thin wrapper.)

We could also have a thin wrapper for `git::clone` which returns the default branch as `git::fetch` doesn't need to. And I'll be able to add `git::sync` that does the extra stuff I need.

If there are no objections, I go ahead and send out a few PRs that do this. 
> 
> @essiene might be interested.

Indeed. Thanks for looping me in.

yuja commented:
> We could also have a thin wrapper for `git::clone` which returns the default branch as `git::fetch` doesn't need to. And I'll be able to add `git::sync` that does the extra stuff I need.

Sounds good, thanks.

BTW, I used `peer` in the example, but it's probably a bad word. I expect `peer` to be associated with a single remote.

essiene commented:
> > We could also have a thin wrapper for `git::clone` which returns the default branch as `git::fetch` doesn't need to. And I'll be able to add `git::sync` that does the extra stuff I need.
> 
> Sounds good, thanks.
> 
> BTW, I used `peer` in the example, but it's probably a bad word. I expect `peer` to be associated with a single remote.

Yeah, makes sense.

essiene commented:
> > > We could also have a thin wrapper for `git::clone` which returns the default branch as `git::fetch` doesn't need to. And I'll be able to add `git::sync` that does the extra stuff I need.
> > 
> > 
> > Sounds good, thanks.
> > BTW, I used `peer` in the example, but it's probably a bad word. I expect `peer` to be associated with a single remote.
> 
> Yeah, makes sense.

#4959 & #4960

mateon1 commented:
Closed due to being superseded by #5498
